SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestReplicatedStartupActor/SpatialTestReplicatedStartupActor.cpp

```diff
@@ -44,36 +56,136 @@ void ASpatialTestReplicatedStartupActor::BeginPlay()
 {
 	Super::BeginPlay();
 
-	AddStep(TEXT("SpatialTestReplicatedStartupActorClientsSetup"), FWorkerDefinition::AllClients, [this](ASpatialFunctionalTest* NetTest)
+	// Common Setup
+
+	// All workers set a reference to the ReplicatedStartupActor.
+	AddStep(TEXT("SpatialTestReplicatedStartupActorUniversalReferenceSetup"), FWorkerDefinition::AllWorkers, nullptr, nullptr, [this](ASpatialFunctionalTest* NetTest, float DeltaTime)
+		{
+			TArray<AActor*> ReplicatedStartupActors;
+			UGameplayStatics::GetAllActorsOfClass(GetWorld(), AReplicatedStartupActor::StaticClass(), ReplicatedStartupActors);
+
+			if (ReplicatedStartupActors.Num() == 1)
+			{
+				ReplicatedStartupActor = Cast<AReplicatedStartupActor>(ReplicatedStartupActors[0]);
+				FinishStep();
+			}
+		}, 5.0f);
+
+	// Phase 1
+
+	// All clients send a server RPC from the ReplicatedStartupActor.
+	AddStep(TEXT("SpatialTestReplicatedStartupActorClientsSendRPC"), FWorkerDefinition::AllClients, nullptr, nullptr, [this](ASpatialFunctionalTest* NetTest, float DeltaTime)
 		{
-			// Make sure that the PlayerController has been set before trying to do anything with it, this might prevent Null Pointer exceptions being thrown when UE ticks at a relatively slow rate
 			AReplicatedStartupActorPlayerController* PlayerController = Cast<AReplicatedStartupActorPlayerController>(GetLocalFlowController()->GetOwner());
-			return IsValid(PlayerController);
-		},
-		[this](ASpatialFunctionalTest* NetTest)
+
+			// Make sure that the PlayerController has been set before trying to do anything with it, this might prevent Null Pointer exceptions being thrown when UE ticks at a relatively slow rate
+			if (IsValid(PlayerController))
+			{
+				PlayerController->ClientToServerRPC(this, ReplicatedStartupActor);
+				FinishStep();
+			}
+		}, 5.0f);
+
+	// All clients check that the RPC was received and correctly applied.
+	AddStep(TEXT("SpatialTestReplicatedStarupActorClientsCheckRPC"), FWorkerDefinition::AllClients, nullptr, nullptr, [this](ASpatialFunctionalTest* NetTest, float DeltaTime)
 		{
-			TArray<AActor*> ReplicatedActors;
-			UGameplayStatics::GetAllActorsOfClass(GetWorld(), AReplicatedTestActorBase::StaticClass(), ReplicatedActors);
+			if (bIsValidReference)
+			{
+				AReplicatedStartupActorPlayerController* PlayerController = Cast<AReplicatedStartupActorPlayerController>(GetLocalFlowController()->GetOwner());
 
-			checkf(ReplicatedActors.Num() == 1, TEXT("There should be exactly 1 replicated actor"));
+				// Reset the variables to allow for relevant consecutive runs of the same test.
+				PlayerController->ResetBoolean(this);
```

Comment:
I think this could be moved to the universal setup? Instead of resetting mid-test

Reply:
Yes that's true, makes more sense to have it in the universal,  thank you for the sugesstion.